### PR TITLE
Editor: UIPoints honors UNDO/REDO  

### PR DESCRIPTION
**Pull request by ycw**
**Description**

With this PR, LatheGeometry's `Points` input and TubeGeometry's `Path` input will honor UNDO/REDO.

- Fixed `UIPoints.clear()` 
  - Deletes point rows one by one from tail
  - Deletes `this.lastPointIdx=0` (redundant) 
- Fixed `UIPoints2.setValue()`, and `UIPoints3.setValue()`
  - Re-generate point rows only if needed

Preview: https://raw.githack.com/ycw/three.js/editor-uipoints-honors-undo-redo/editor/index.html

### CODE REVIEW

File: editor/js/Sidebar.Geometry.LatheGeometry.js [hunk 1]
```diff
@@ -57,10 +57,37 @@ function GeometryParametersPanel( editor, object ) {
...
+
+		return true;
+
+	}
+
 	function refreshUI() {
 
 		const parameters = object.geometry.parameters;
 
+		if ( isPointsIdentical( parameters.points ) === false ) {
```
Comment (Mugen87) [inline editor/js/Sidebar.Geometry.LatheGeometry.js:83]: How about implementing the equals checks in `setValue()` instead`? In this way, you can just execute `points.setValue( parameters.points );` in `refreshUI()`.

In `setValue()`, the UI is only build when the points set is different. This makes it also possible to control whether `update()` is called or not. So you don't have to change the `onChange()` handler.

File: editor/js/libs/ui.three.js [hunk 2]
```diff
@@ -571,16 +571,11 @@ class UIPoints extends UISpan {
 
 	clear() {
 
-		for ( let i = 0; i < this.pointsUI.length; ++ i ) {
```
Comment (Mugen87) [inline editor/js/libs/ui.three.js:574]: Instead of using a while loop, can you also reverse the order in the `for` loop instead. 
```js
for ( let i = this.pointsUI.length - 1; i >= 0; -- i ) {
```

Comment (ycw) [reply editor/js/libs/ui.three.js:574]: ok.

File: editor/js/Sidebar.Geometry.LatheGeometry.js (hunk 1 above)
Comment (ycw) [reply editor/js/Sidebar.Geometry.LatheGeometry.js]: ok.

File: editor/js/libs/ui.three.js [hunk 3]
```diff
@@ -669,7 +684,6 @@ class UIPoints2 extends UIPoints {
 
 		}
 
-		this.update();
```
Comment (Mugen87) [inline editor/js/libs/ui.three.js:672]: Why did you delete this line?

Comment (Mugen87) [inline editor/js/libs/ui.three.js:672]: IMO, it's more correct to keep this line. The update callback should be executed whenever the state of the UI element changes. That is when values change.

Comment (ycw) [reply editor/js/libs/ui.three.js:672]: then we need this again ...

```js
points.onChange( null )
points.setValue( .. )
points.onChange( update )
```
... to avoid circular

Comment (ycw) [reply editor/js/libs/ui.three.js]: ok

File: editor/js/libs/ui.three.js [hunk 4]
```diff
@@ -660,6 +673,8 @@ class UIPoints2 extends UIPoints {
 
 	setValue( points ) {
 
+		if ( this.isPointsIdentical( points ) === true ) return this;
```
Comment (Mugen87) [inline editor/js/libs/ui.three.js:676]: When thinking about this we should remove this bit since it's over-engineering. The overhead of generating the UI is so minor that we don't have to test for equal values.

Comment (ycw) [reply editor/js/libs/ui.three.js:676]: This identical checking here is to distinguish command path from regular path, and then decides that should point rows be re-generated or not. Blindly re-generating point rows will cause bug in regular path (see below).

Command path: The points value won't be identical there:

```
undo/execute SetGeometryCommand  # geometry is updated
   dispatch geometryChanged
      call refreshUI()
          call points.setValue(geometry.parameter.path.points # << not identical to points.getValue()
```
In this case, re-generating point rows in UIPoints is needed because UI out-synced its data.

Regular path: The points will be identical there:

```
on number input changed
  call update()
    execute SetGeometryCommand
      dispath geometryChanged
        call refreshUI()
          call points.setValue(geometry.parameter.path.points # << identical to points.getValue()
```

In this case, **re-generating point rows in UIPoints is not needed** because UI already synced its data, **otherwise, (BUG) users will slide a ghost input** (users are sliding an old input while UI is showing the new one): 

https://github.com/mrdoob/three.js/assets/1063018/fddc96a8-881d-4608-a830-e4e263a531d3

To conclude, this identical checking is needed, not just for perf, but also avoiding abovementioned bug.

Comment (Mugen87) [reply editor/js/libs/ui.three.js:676]: Um, let me think about this. Maybe there is a different solution that allows to implement use to implement something more simple.

**Comment by Mugen87** [comment]
Closing in favor of #28677.
